src/cpp/session/modules/data/DataViewer.cpp

```diff
@@ -1080,26 +1058,56 @@ void onDetectChanges(module_context::ChangeSource source)
          // clear working data for the object
          r::exec::RFunction(".rs.removeWorkingData", i->first).call();
    
-         if (sexp != nullptr && Rf_inherits(sexp, "data.frame"))
+         if (sexp != nullptr)
          {
             // create a new frame object to capture the new state of the frame
             CachedFrame newFrame(i->second.envName, i->second.objName, sexp);
-
-            // replace cached copy
-            r::exec::RFunction(".rs.assignCachedData", 
-                  i->first, sexp, i->second.objName).call();
-
-            // emit client event
-            json::Object changed;
-            changed["cache_key"] = i->first;
-            changed["structure_changed"] = i->second.ncol != newFrame.ncol || 
-               i->second.colNames != newFrame.colNames;
-            changed["type_changed"] = false;
-            ClientEvent event(client_events::kDataViewChanged, changed);
-            module_context::enqueClientEvent(event);
-
-            // replace old frame with new
-            s_cachedFrames[i->first] = newFrame;
+            
+            // check for changes in the object
+            bool typeChanged = false;
+            if (i->second.observedSEXP != nullptr)
+            {
+               SEXP oldClass = Rf_getAttrib(i->second.observedSEXP, R_ClassSymbol);
+               SEXP newClass = Rf_getAttrib(sexp, R_ClassSymbol);
+               typeChanged = !R_compute_identical(oldClass, newClass, 0);
+            }
+            
+            bool structureChanged =
+                  i->second.nrow != newFrame.nrow ||
+                  i->second.ncol != newFrame.ncol ||
+                  i->second.colNames != newFrame.colNames;
+            
+            if (typeChanged || structureChanged)
+            {
+               // replace cached copy
+               r::exec::RFunction(".rs.assignCachedData")
+                     .addParam(i->first)
+                     .addParam(sexp)
+                     .addParam(i->second.objName)
+                     .call();
+               
+               // figure out the object classes for this thing
+               std::vector<std::string> objectClass;
+               Error error = r::exec::RFunction("base:::class")
+                     .addParam(sexp)
+                     .call(&objectClass);
+               
+               if (error)
+                  LOG_ERROR(error);
+
+               // emit client event
+               json::Object changed;
+               changed["cache_key"] = i->first;
+               changed["type_changed"] = typeChanged;
+               changed["structure_changed"] = structureChanged;
+               changed["object_exists"] = true;
+               changed["object_class"] = json::toJsonArray(objectClass);
```

Comment:
Not currently used, but figure it could be useful if we later want to try and transition the view from Data Viewer to Object Explorer.